### PR TITLE
AMDGPU: Fix using illegal VOP3 literal in frame index elimination

### PR DESCRIPTION
**Pull request by arsenm**
Most of the change is just moving the default case outside
of the switch so we can break to the default handling for arbitrary
operations.

### CODE REVIEW

**Comment by arsenm** [comment]
* **#115747** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/115747?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by arsenm** [comment]
ping

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp [hunk 1]
```diff
@@ -2496,6 +2496,25 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
...
+      if (OtherOp->isImm()) {
+        int64_t TotalOffset = OtherOp->getImm() + Offset;
+
+        if (!ST.hasVOP3Literal() && SIInstrInfo::isVOP3(*MI) &&
+            !AMDGPU::isInlinableIntLiteral(TotalOffset)) {
+          // If we can't support a VOP3 literal in the VALU instruction, we
+          // can't specially fold into the add.
+          // TODO: Handle VOP3->VOP2 shrink to support the fold.
+          break;
+        }
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2512]: I assume this is the only significant change?

Why does this block have to move before the !enableFlatScratch handling?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2512]: To ensure it doesn't call the scavenger if this path isn't going to handle the case

File: llvm/test/CodeGen/AMDGPU/local-stack-alloc-block-sp-reference.ll [hunk 2]
```diff
@@ -20,8 +20,9 @@ define amdgpu_kernel void @local_stack_offset_uses_sp(ptr addrspace(1) %out) {
 ; MUBUF-LABEL: local_stack_offset_uses_sp:
 ; MUBUF:       ; %bb.0: ; %entry
 ; MUBUF-NEXT:    s_add_u32 s0, s0, s17
+; MUBUF-NEXT:    v_mov_b32_e32 v1, 0x3000
 ; MUBUF-NEXT:    s_addc_u32 s1, s1, 0
-; MUBUF-NEXT:    v_mov_b32_e32 v0, 0x3040
+; MUBUF-NEXT:    v_add_u32_e32 v0, 64, v1
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/local-stack-alloc-block-sp-reference.ll:25]: Would your "TODO: Handle VOP3->VOP2 shrink to support the fold" fix this regression?

**Review by jayfoad — APPROVED** [review]
LGTM.

**Pull request merged**